### PR TITLE
Add demo of fetching data from the Community API (nDelius)

### PR DESCRIPTION
**Pull request by Gweaton**
Note: credit where credit's due - a lot of this PR was inspired by/based on @lawrence-forooghian's [hard work](https://github.com/ministryofjustice/hmpps-interventions-add-find-prototype/pull/17) to integrate with the Offender Assessment API.

## What's new

We were hoping that the [Community API](https://github.com/ministryofjustice/community-api) would be a good data source for a lot of the personal information about a service user, e.g. name, address, DoB, contact details. This PR adds a working demo to show that it's possible to fetch data about a test user from a locally-running instance of the Community API (and hopefully in future a staging/production version), due to limitations around getting VPN access required to test with the real thing.

This allows users to search for an existing service user by CRN (an existing user in the local data set has a CRN of X320741) and view that user's personal information. **Unfortunately not all of the data we expect from the Community API is present on this example service user e.g. other aliases, ethnicity, sexual orientation, disability information, but this will have to do until we get access to the VPN when we'll have access to a more complete dataset.**

Running this search requires running a local version of the [Community API](https://github.com/ministryofjustice/community-api). You can run this (and the [HMPPS Auth](https://github.com/ministryofjustice/hmpps-auth) service required for generating an authentication token) from that project by running `docker-compose up`.

### Search for a service user by CRN
image

### View example data about that user
image

## Next steps

- Once we have VPN access we should verify that the Community API returns a more complete data set for a user, as I'm unsure if there are any more complete users in the local version. A current analysis of the fields returned by the API was written up by Chris Heathcote and is available in [this sheet](https://docs.google.com/spreadsheets/d/1-YrQlcGoffKdn8KnoHnvHzP28pA28Ymr5KZdjrenyL0/edit?ts=5f58b66d#gid=1619951312) (you may need to request access).

### CODE REVIEW

File: lib/hmpps_oauth_client.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+exports.getHMPPSAuthToken = async function () {
+  try {
+    const response = await axios.post(
+      "http://localhost:8090/auth/oauth/token?grant_type=client_credentials&scope=",
+      {},
+      {
+        headers: {
+          Authorization: `Basic ${Buffer.from(
+            "prisonstaffhubclient:clientsecret",
```
Comment (sldblog) [inline lib/hmpps_oauth_client.js:11]: Is `prisonstaffhubclient` intended at this point?

File: lib/community_api_client.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const axios = require("axios");
+
+exports.getServiceUserDetailsByCRN = async function (crn, authToken) {
+  try {
+    const response = await axios.get(
+      `http://localhost:8096/secure/offenders/crn/${crn}/all`,
```
Comment (sldblog) [inline lib/community_api_client.js:6]: What do you think about adding a section in the readme about booting these services up on these ports? (Just in case we'll have to boot it up in a hurry)

File: app/views/index.html [hunk 3]
```diff
@@ -18,28 +18,34 @@
...
+            </svg>
+          </a>
+          <a href="book-and-manage/index" role="button" draggable="false" class="govuk-button govuk-button--start govuk-!-margin-top-2 govuk-!-margin-bottom-8" data-module="govuk-button">
+            Book and manage an intervention
+            <svg class="govuk-button__start-icon" xmlns="http://www.w3.org/2000/svg" width="17.5" height="19" viewbox="0 0 33 40" role="presentation" focusable="false">
+              <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z"></path>
+            </svg>
+          </a>
+          <a href="service-user-details/search" role="button" draggable="false" class="govuk-button govuk-button--start govuk-!-margin-top-2 govuk-!-margin-bottom-8" data-module="govuk-button">
+            Look up a service user
```
Comment (sldblog) [inline app/views/index.html:41]: Should we mention the source (given that we have 3 big systems with service user data 😂)?
```suggestion
            Look up a service user from nDelius via community-api
```

File: app/views/service-user-details/search.html [hunk 4]
```diff
@@ -0,0 +1,55 @@
+{% extends "layout.html" %}
+
+{% block pageTitle %}
+  MoJ Interventions - make referrals
```
Comment (sldblog) [inline app/views/service-user-details/search.html:4]: Should this be "make referrals" at this point?

File: app/views/service-user-details/search.html [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    <div class="govuk-grid-column-two-thirds">
+      <form action="/service-user-details/show" class="form">
+        {{ govukInput({
+          label: {
+            text: "Enter service user CRN",
+            classes: "govuk-label--xl",
+            isPageHeading: true
+          },
+          hint: {
+            text: "Searching for a user requires a running instance of the Community API"
```
Comment (sldblog) [inline app/views/service-user-details/search.html:36]: Our users may not know about the Community API -- what do you think about mentioning nDelius instead?

**Comment by sldblog** [comment]
📝 There's an alternative service for _searching_ for service users called probation offender search ([APIdocs](https://probation-offender-search-dev.hmpps.service.justice.gov.uk/swagger-ui.html)). It's an [ElasticSearch-based solution](https://github.com/ministryofjustice/probation-offender-search) which makes instant searches possible.

The API-backed searching may have unexpected performance issues on production (as it's using the nDelius database directly for searching).

Basically, I wonder if it'd be worth having a copy of this demo with the ElasticSearch one too?

File: lib/hmpps_oauth_client.js (hunk 1 above)
Comment (Gweaton) [reply lib/hmpps_oauth_client.js:11]: I had a conversation with Andy Marke from the Community API team, and he sent me [this base64 encoded string](https://mojdt.slack.com/archives/CR5ESQ8T1/p1599723574056300?thread_ts=1599665934.052800&cid=CR5ESQ8T1) which contains `prisonstaffhubclient`. I felt it was better to put in the credentials themselves as it's (possibly?) a bit clearer.

File: lib/community_api_client.js (hunk 2 above)
Comment (Gweaton) [reply lib/community_api_client.js:6]: Sounds good - I put it in the commit message, but our README could definitely do with some love.

File: app/views/service-user-details/search.html (hunk 5 above)
Comment (Gweaton) [reply app/views/service-user-details/search.html]: Have added this to the page title.

**Comment by Gweaton** [comment]
> 📝 There's an alternative service for _searching_ for service users called probation offender search ([APIdocs](https://probation-offender-search-dev.hmpps.service.justice.gov.uk/swagger-ui.html)). It's an [ElasticSearch-based solution](https://github.com/ministryofjustice/probation-offender-search) which makes instant searches possible.
> 
> The API-backed searching may have unexpected performance issues on production (as it's using the nDelius database directly for searching).
> 
> Basically, I wonder if it'd be worth having a copy of this demo with the ElasticSearch one too?

This was going to be the next step this week, but now we're going to be building prototypes instead, so I'll leave this for next sprint.
